### PR TITLE
Persistence for RequestConversations

### LINKED ISSUE #37: Implement persistent auditing logs

### PR DESCRIPTION
**Pull request by elbart**
* Closes #37

### CODE REVIEW

**Comment by amotl** [comment]
Good stuff! 🎉

File: fh-http/src/manager/request_conversation.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    Response(fh_v8::response::Response),
+    Log(String),
+}
+
+pub(crate) struct RequestConversation {
+    id: Uuid,
+    created_at: chrono::DateTime<Utc>,
+    request_processor_id: Uuid,
+    audit_entries: Vec<AuditEntry>,
+}
```
Comment (amotl) [inline fh-http/src/manager/request_conversation.rs]: This is a nice data model. Let me just share some thoughts.

As far as the current design of the whole machinery goes, we will have a single main request/response conversation. That is, the ingress HTTP request into the public endpoint and the final response made with `respond_with` (#20).

On the other hand, during the execution of the recipe, different request/response conversations can be made on behalf of `dispatch_request` (#25), right?

So, we might want to discriminate between both variants here. I am proposing a `main` slot for the embracing HTTP conversation and an `items` slot for the inner HTTP conversations.

In order to give immediate feedback to the user, we want to fill in the slots early throughout the execution of the recipe. That is:

1. Request comes in.
2. Create `RequestConversation` instance.
3. Fill in `main.request`.
4. On `dispatch_request`, fill in `items[0].request`.
5. When receiving the response, fill in `items[0].response`.
6. Optionally repeat 4. and 5. for more invocations of `dispatch_request`.
7. Before or after finishing the recipe through an explicit or implict `respond_with`, fill in `main.response`.

What do you think about this?

```suggestion
pub(crate) struct RequestConversation {
    id: Uuid,
    created_at: chrono::DateTime<Utc>,
    request_processor_id: Uuid,
    main: AuditEntry,
    items: Vec<AuditEntry>,
}
```

P.S.: Shouldn't `AuditEntry` also be a `struct` to have the convenience of named attributes in there?

Comment (elbart) [reply fh-http/src/manager/request_conversation.rs]: 1. Your data structure suggestion: Well currently an `AuditEntry` is an enum which can have one variant at a time, containing it's respective data / types. So you can't put a Request and Response object in an AuditEntry together but I see the point that these should be couled together. Unfortunately, enum variants are not first-class types, yet [1] so we cannot ensure that an Item has a specific variant just by TypeDefinitions. However I do not want to loose the logging functionality. The `conversation_audit_log` table should be generic enough to contain request, responses, log entries and possibly more. I will look into how to couple these together. Thanks for pointing that out. Also, I agree about the "store something as soon as we can" approach, which is exactly what I wanted to achieve with the design.

2. Your PS: We can have named attributes in the enum variants, too - rust offers that and defines it as idiomatic. So i can change that, but its fone to do that later when we need it

[1]: https://github.com/rust-lang/rfcs/pull/2593

Comment (elbart) [reply fh-http/src/manager/request_conversation.rs]: For now, I kept the design a bit simpler. We only store a Vec<AuditItem> in the `RequestConversation` struct. For now we have a convention, that the first entry is always the incoming request. I implemented a `parent` pattern, which we can use to link responses to the incoming request or self-issued requests. This way we couple Requests and Responses together. That's all I can do so far, here.

**Pull request merged**